docs/observability.asciidoc

```diff
@@ -0,0 +1,251 @@
+client.on('response', (err, req) => {
+  console.log(err, req)
+})
+----
+
+|`sniff`
+a|Emitted when the clients ends a sniffing request.
+[source,js]
+----
+client.on('sniff', (err, req) => {
+  console.log(err, req)
+})
+----
+
+|`resurrect`
+a|Emitted if the client is able to resurrect a dead node.
+[source,js]
+----
+client.on('resurrect', (err, req) => {
+  console.log(err, req)
+})
+----
+
+|===
+
+The values of `req` in `request`, `response` and `sniff` will be:
+[source,ts]
+----
+  body: any;
+  statusCode: number | null;
+  headers: anyObject | null;
+  warnings: string[] | null;
+  meta: {
+    context: any;
+    name: string;
+    request: {
+      params: TransportRequestParams;
+      options: TransportRequestOptions;
+      id: any;
+    };
+    connection: Connection;
+    attempts: number;
+    aborted: boolean;
+    sniff?: {
+      hosts: any[];
+      reason: string;
+    };
+  };
+----
+
+While the `req` value in `resurrect` will be:
+[source,ts]
+----
+export interface ResurrectEvent {
+  strategy: string;
+  isAlive: boolean;
+  connection: Connection;
+  name: string;
+  request: {
+    id: any;
+  };
+}
+----
+
+=== Correlation id
+Correlate events can be quite hard, especially if there are happening many events at the same time, the client offers you an automatic (and configurable) system to help you handle this problem.
+[source,js]
+----
+const { Client } = require('@elastic/elasticsearch')
+const client = new Client({ node: 'http://localhost:9200' })
+
+client.on('request', (err, event) => {
+  const { id } = event.meta.request
+  if (err) {
+    console.log({ error: err, reqId: id })
+  }
+})
+
+client.on('response', (err, event) => {
+  const { id } = event.meta.request
+  if (err) {
+    console.log({ error: err, reqId: id })
+  }
+})
+
+client.search({
+  index: 'my-index',
+  body: { foo: 'bar' }
+}, (err, result) => {
+  if (err) console.log(err)
+})
+----
+
+By default the id is an incremental integer, but you can easily configure that with the `generateRequestId` option:
+[source,js]
+----
+const { Client } = require('@elastic/elasticsearch')
+const client = new Client({
+  node: 'http://localhost:9200',
+  generateRequestId: function () {
+    // your id generation logic
+    // must be syncronous
+    return 'id'
+  }
+})
+----
+
+You can also specify a custom id per request:
+[source,js]
+----
+client.search({
+  index: 'my-index',
+  body: { foo: 'bar' }
+}, {
+  id: 'custom-id'
+}, (err, result) => {
+  if (err) console.log(err)
+})
+----
+
+=== Context object
+Sometimes, you might need to make some custom data available in your events, you can do that via the `context` option of a request: 
+[source,js]
+----
+const { Client } = require('@elastic/elasticsearch')
+const client = new Client({ node: 'http://localhost:9200' })
+
+client.on('request', (err, event) => {
+  const { id } = event.meta.request
+  const { context } = event.meta
+  if (err) {
+    console.log({ error: err, reqId: id, context })
```

Comment:
I wonder if providing the winter key would make this example more clear, though they would obviously want to protect that it exists on the context before accessing like I did:

```
client.on('request', (err, event) => {
  const { id } = event.meta.request
  const { winter } = event.meta.context
  if (err) {
    console.log({ error: err, reqId: id, winter })
  }
})
```

Reply:
Good point!